### PR TITLE
Add '--no-fail' option

### LINKED ISSUE #391: Suppress Exit Code
- [ ] This is a bug report
- [x] This is a feature request
- [x] I searched existing issues before opening this one

### Expected behavior

When running in a CICD environment, I would like to suppress the exit code on findings. That way I would know when the command fails due to wrong usage (file missing) and could evaluate the results that are written to a checkstyle result file later on in Jenkins / SonarQube. 
I would like to pass something like `--suppress-exit-code`

### Actual behavior

Exit code is `1` if I have finding and command execution issues.

### PR DESCRIPTION
**Pull request by antechrestos**
Each formatter shall exit non 0 code when any error occurs unless '--no-fail' option specified. If so, a 0 status code is returned even if any rule is violated

Closes #391

### What I did

Make program return non zero code if failure occurs, whenever the formatter. If `--no-fail` is specified , the programm returns `0` status code, whatever rules are violated or not.

### How I did it

- remove treatment of exit in lint and put it in main
- added option
- print result and analyse the exit method that shall be called

### How to verify it

- launch binary with invalid `Dockerfile` with `-f gitlab_codeclimate` and expects `1`status code
- launch binary with invalid `Dockerfile` with `-f gitlab_codeclimate` and `--no-fail` and expects `0`status code
- launch binary with valid `Dockerfile` with `-f gitlab_codeclimate` and expects `0`status code
- launch binary with valid `Dockerfile` with `-f gitlab_codeclimate` and `--no-fail` and expects `0`status code

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -6,8 +6,10 @@ module Main where
 
 import Control.Applicative
 import qualified Data.List.NonEmpty as NonEmpty
-import Data.Semigroup ((<>))
+import Data.Semigroup()
```
Comment (lorenzo) [inline app/Main.hs:9]: Please leave the line as it was, it is meant for older versions of haskell

File: app/Main.hs [hunk 2]
```diff
@@ -6,8 +6,10 @@ module Main where
...
 import qualified Data.Set as Set
+import qualified Data.Sequence as Seq
+import Data.List.NonEmpty()
```
Comment (lorenzo) [inline app/Main.hs:12]: please use `qualified` and remove the parenthesis

File: app/Main.hs [hunk 3]
```diff
@@ -68,13 +71,16 @@ parseOptions :: Parser CommandOptions
...
     <$> version -- CLI options parser definition
+    <*> noFail
     <*> configFile
     <*> outputFormat
     <*> files
     <*> lintOptions
   where
     version = switch (long "version" <> short 'v' <> help "Show version")
 
+    noFail = switch (long "no-fail" <> help "Do not fail when any rule is violated")
```
Comment (lorenzo) [inline app/Main.hs:82]: `Don't exit with a failure status code when any rule is violated`

File: src/Hadolint/Lint.hs [hunk 4]
```diff
@@ -38,21 +37,15 @@ data OutputFormat
...
-        Codacy -> Codacy.printResult res >> exitSuccess
+printResults :: OutputFormat -> Format.Result Text DockerfileError -> IO ()
+printResults format allResults = do
+  case format of
+    TTY -> TTY.printResult allResults
+    Json -> Json.printResult allResults
+    Checkstyle -> Checkstyle.printResult allResults
+    CodeclimateJson -> Codeclimate.printResult allResults
+    GitlabCodeclimateJson -> Codeclimate.printGitlabResult allResults
+    Codacy -> Codacy.printResult allResults
```
[the remark this answers is not in the document]
Comment (antechrestos) [reply src/Hadolint/Lint.hs:48]: @lorenzo can I knwo why? The code does not brings any info on it.

**Comment by antechrestos** [comment]
@lorenzo is the implementation ok? Shall I need to change anything? Please let me know

**Comment by lorenzo** [comment]
Thanks, it's great now :)

**Comment by antechrestos** [comment]
Thanks @lorenzo

It's Sunday you should rest ☺

**Pull request merged**